### PR TITLE
Introduce ScriptGraph to calculate script execution order

### PR DESCRIPTION
**Pull request by turbo1912**

### CODE REVIEW

File: src/fal/dag.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+    def parse(self) -> List[str]:
+        """
+        Parses the python file to identify any dependencies
+        """
+        with open(self.path) as file:
+            lines = file.readlines()
+            ## TODO: do this search in parallel
+            dependency_names = []
+            for line in lines:
```
Comment (chamini2) [inline src/fal/dag.py:30]: Seems to be very easy with the `with` approach: https://docs.python.org/3/library/multiprocessing.html#introduction

File: src/fal/cli.py [hunk 2]
```diff
@@ -36,5 +37,8 @@ def run(run, project_dir, profiles_dir, keyword, all):
     real_profiles_dir = os.path.realpath(os.path.normpath(profiles_dir))
 
     project = parse_project(real_project_dir, real_profiles_dir, keyword)
-    for model in project.get_filtered_models(all):
+    models = project.get_filtered_models(all)
+    ordered_scripts = ScriptGraph(models, keyword, project_dir).sort()
+    ## TODO: Run ordered scripts instead
```
Comment (chamini2) [inline src/fal/cli.py:42]: Why are we not using it in this PR?

File: src/fal/dag.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+class ScriptGraphBuilder:
+    modelNameToModelLookup: Dict[str, DbtModel] = {}
+    modelToScriptLookup: Dict[str, List[Path]] = {}
+    falScripts: Dict[UniqueKey, FalScript] = {}
+
+    def __init__(self, models: List[DbtModel], keyword: str, root: str):
+        for model in models:
+            self.modelToScriptLookup[model.name] = model.get_scipts(keyword, root)
+            self.modelNameToModelLookup[model.name] = model
+        # self.scripts = _flatten(self.modelToScriptLookup.values())
```
Comment (chamini2) [inline src/fal/dag.py:70]: delete?

File: src/fal/dag.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            )
+
+@dataclass(frozen=True)
+class UniqueKey:
+    model: DbtModel
+    script_path: Path
+
+class ScriptGraphBuilder:
+    modelNameToModelLookup: Dict[str, DbtModel] = {}
+    modelToScriptLookup: Dict[str, List[Path]] = {}
```
Comment (chamini2) [inline src/fal/dag.py:63]: `modelNameToModelLookup` and `modelToScriptLookup` could be a single lookup map with tuples. Not that it is necessarily better.

File: src/fal/dag.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            self.modelToScriptLookup[model.name] = model.get_scipts(keyword, root)
+            self.modelNameToModelLookup[model.name] = model
+        # self.scripts = _flatten(self.modelToScriptLookup.values())
+        for model_name, script_list in self.modelToScriptLookup.items():
+            for script in script_list:
+                self.recursivelySetDependencies(UniqueKey(self.modelNameToModelLookup[model_name], script))
+
+    def recursivelySetDependencies(self, key: UniqueKey):
+        if (self.falScripts.get(key) is None):
+            self.falScripts[key] = FalScript(key.model, key.script_path)
```
Comment (chamini2) [inline src/fal/dag.py:77]: What if we change this first few lines to check if it is already set and return right away? I think once it is set, we don't want to continue with the rest of the script, right?

```py
if key in self.falScripts:
  return

current_script = FalScript(...)
self.falScripts[key] = current_script
```

The interesting thing about that is that then inside the `for`, in line 87 we no longer have to check before calling. We can call `self.recursivelySetDependencies(UniqueKey(model, dependency_scipt))` and it checks for us in the top of the method.

File: src/fal/dag.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        for model in models:
+            self.modelToScriptLookup[model.name] = model.get_scipts(keyword, root)
+            self.modelNameToModelLookup[model.name] = model
+        # self.scripts = _flatten(self.modelToScriptLookup.values())
+        for model_name, script_list in self.modelToScriptLookup.items():
+            for script in script_list:
+                self.recursivelySetDependencies(UniqueKey(self.modelNameToModelLookup[model_name], script))
+
+    def recursivelySetDependencies(self, key: UniqueKey):
+        if (self.falScripts.get(key) is None):
```
Comment (chamini2) [inline src/fal/dag.py:76]: Can use syntax like `if key not in self.falScripts:` or `if key in self.falScripts`

File: src/fal/cli.py [hunk 7]
```diff
@@ -1,5 +1,6 @@
 import click
 import os
+from fal.dag import ScriptGraph, ScriptGraphBuilder
```
Comment (chamini2) [inline src/fal/cli.py:3]: not using ScriptGraphBuilder

File: src/fal/dag.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    def __init__(self, models: List[DbtModel], keyword: str, root: str):
+        for model in models:
+            self.modelToScriptLookup[model.name] = model.get_scipts(keyword, root)
+            self.modelNameToModelLookup[model.name] = model
+        # self.scripts = _flatten(self.modelToScriptLookup.values())
+        for model_name, script_list in self.modelToScriptLookup.items():
+            for script in script_list:
+                self.recursivelySetDependencies(UniqueKey(self.modelNameToModelLookup[model_name], script))
+
+    def recursivelySetDependencies(self, key: UniqueKey):
```
Comment (chamini2) [inline src/fal/dag.py:75]: snake case?

File: src/fal/dag.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
+from dataclasses import dataclass, field
```
Comment (chamini2) [inline src/fal/dag.py:1]: Maybe we could write some tests for this file?

Comment (chamini2) [reply src/fal/dag.py:1]: Not specific ones, just passing in examples and checking that the results make sense.

File: src/fal/dag.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        dependency_models = current_script.parse()
+       
+        for model_name in dependency_models:
+            model = self.modelNameToModelLookup[model_name]
+            dependency_scripts = self.modelToScriptLookup.get(model_name, [])
+
+            for dependency_scipt in dependency_scripts:
+                if (self.falScripts.get(UniqueKey(model, dependency_scipt)) is None):
+                    self.recursivelySetDependencies(UniqueKey(model, dependency_scipt))
+                current_script.addDependency(self.falScripts.get(UniqueKey(model, dependency_scipt)))
```
Comment (chamini2) [inline src/fal/dag.py:89]: you build `UniqueKey(model, dependency_scipt)` 3 times, maybe create it in line 87 as a variable inside the for loop and reuse it?

Comment (chamini2) [reply src/fal/dag.py:89]: Also, `_scipt` -> `_script`

File: src/fal/dag.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    
+    def getValues(self) -> List[FalScript]:
+        return self.falScripts.values()
+
+
+class ScriptGraph:
```
Comment (chamini2) [inline src/fal/dag.py:95]: Still have to review this class more

File: src/faldbt/project.py [hunk 12]
```diff
@@ -24,9 +25,15 @@ def __post_init__(self):
         self.meta = self.node.config.meta
         self.columns = self.node.columns
 
-    def model_key(self, project_name):
+    def __hash__(self) -> int:
+        return self.name.__hash__()
+        
+    def model_key(self, project_name) -> str:
         return "model." + project_name + "." + self.name
```
Comment (chamini2) [inline src/faldbt/project.py:32]: use `self.model.unique_id` and maybe reuse that for the hash?

File: src/fal/dag.py (hunk 3 above)
Comment (turbo1912) [reply src/fal/dag.py:70]: yeaa

File: src/fal/dag.py (hunk 11 above)
Comment (turbo1912) [reply src/fal/dag.py:95]: I am using this algorithm by the way, helps with the review 

https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm

File: src/fal/dag.py (hunk 1 above)
Comment (turbo1912) [reply src/fal/dag.py:30]: if its ok with you i am going to delay this as a future task..

File: src/fal/dag.py (hunk 10 above)
Comment (turbo1912) [reply src/fal/dag.py:89]: done

File: src/fal/dag.py (hunk 9 above)
Comment (turbo1912) [reply src/fal/dag.py:1]: yea we should

File: src/fal/dag.py (hunk 8 above)
Comment (chamini2) [reply src/fal/dag.py]: This applies to several methods @turbo1912

File: src/fal/dag.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        dependency_models = current_script.parse()
+       
+        for model_name in dependency_models:
+            model = self.modelNameToModelLookup[model_name]
+            dependency_scripts = self.modelToScriptLookup.get(model_name, [])
+
+            for dependency_script in dependency_scripts:
+                dependency_key = UniqueKey(model, dependency_script)
+                if (self.falScripts.get(dependency_key) is None):
+                    self.recursively_set_dependencies(dependency_key)
```
Comment (chamini2) [inline src/fal/dag.py:88]: You can remove the `if` and call directly. It will be handled by the line 75 when called

File: src/fal/dag.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+                self.outgoing[edge] = []
+        
+
+    def sort(self):
+        """
+        Topologically sorts the python scripts to make sure all dependencies
+        run before the script itsef
+        """
+        leaf_nodes = list(filter(lambda key: len(self.incoming[key]) == 0, self.incoming.keys()))
+        while (len(leaf_nodes) != 0):
```
Comment (chamini2) [inline src/fal/dag.py:118]: `while leaf_nodes:`

File: src/fal/dag.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+            node = leaf_nodes.pop(0)
+            self.ordered_list.append(node)
+            outgoing_copy = self.outgoing[node].copy()
+            for item in outgoing_copy:
+                self.incoming[item].remove(node)
+                self.outgoing[node].remove(item)
+                if (len(self.incoming[item]) == 0):
```
Comment (chamini2) [inline src/fal/dag.py:125]: `if not self.incoming[item]:`

File: src/fal/dag.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+                    leaf_nodes.append(item)
+        return self.ordered_list
```
Comment (chamini2) [inline src/fal/dag.py:127]: Should we check for cycles?

```
if graph has edges then
    return error   (graph has at least one cycle)
else 
    return L   (a topologically sorted order)
```

**Review by chamini2 — APPROVED** [review]

**Pull request merged**